common/canonical_serialization/src/simple_serializer.rs

```diff
@@ -58,29 +58,33 @@ where
         Ok(self)
     }
 
-    fn encode_btreemap<K: CanonicalSerialize, V: CanonicalSerialize>(
+    fn encode_tuple_iterator<K: CanonicalSerialize, V: CanonicalSerialize, I>(
         &mut self,
-        v: &BTreeMap<K, V>,
-    ) -> Result<&mut Self> {
+        iter: I,
+    ) -> Result<&mut Self>
+    where
+        I: Iterator<Item = (K, V)>,
+    {
+        let mut map = BTreeMap::new();
+        for (key, value) in iter {
+            map.insert(
+                SimpleSerializer::<Vec<u8>>::serialize(&key)?,
+                SimpleSerializer::<Vec<u8>>::serialize(&value)?,
+            );
+        }
+
         ensure!(
-            v.len() <= ARRAY_MAX_LENGTH,
+            map.len() <= ARRAY_MAX_LENGTH,
             "array length exceeded the maximum limit. length: {}, max length limit: {}",
-            v.len(),
+            map.len(),
             ARRAY_MAX_LENGTH,
         );
 
         // add the number of pairs in the map
-        self.encode_u32(v.len() as u32)?;
+        self.encode_u32(map.len() as u32)?;
 
         // Regardless of the order defined for K of the map, write in the order of the lexicographic
```

Comment:
I think these belong before line 68

Reply:
We can't do this unfortunately. The problem here is that iterator doesn't have a size. Size can only be derived when we finished reconstructing the map.